### DISCUSSION TITLE
2.0.0

### DISCUSSION THREAD

**Discussion by benhillis** [microsoft/WSL#10487] [General]
*  Added support for new opt-in experimental features to `%USERPROFILE%\.wslconfig`. For full info see the WSL docs update: https://aka.ms/wslsep2023updatedocspage
    * experimental.autoMemoryReclaim - Makes the WSL VM shrink in memory as you use it by reclaiming cached memory
    * experimental.networkingMode - A new networking mode for WSL that adds new features and improves network compatibility
    * experimental.dnsTunneling- Changes how WSL resolves DNS requests to improve network compatibility
    * experimental.firewall- Applies Windows firewall rules to WSL, and allows for advanced firewall controls for the WSL VM
    * experimental.autoProxy- Makes WSL automatically use the proxy information from Windows to improve network compatibility
```
[experimental]
autoMemoryReclaim=gradual | dropcache | disabled
networkingMode=mirrored
dnsTunneling=true
firewall=true
autoProxy=true
```
*  Store WSL is now accessible from Session 0 (solves #9231)
*  All binaries shipped in the WSL package are now signed
*  Introduce `experimental.autoMemoryReclaim` setting to `%USERPROFILE%\.wslconfig`
*  Prompt the user to enable VMP if HNS isn't available on the machine
*  Fix localhost relay failing if ipv6 is disabled
*  Improve error message for WSL1 disabled policy to provide upgrade instructions
*  Fix microk8s scenario for mirrored networking
*  Create shared mounts when init sets up the distro namespace (solves #10410)
*  Detach user block devices before terminating the VM
*  Protect '.wslconfig' from being incorrectly translated
*  Add logic to generate a systemd unit to protect the binfmt interpreter (solves #8843)
*  Fix potential wslservice.exe crash.
*  Retry plan9 mounts on IO error
*  Update Microsoft.WSL.Kernel version 5.15.123.1-1
*  Display a notification prompting the user to install missing optional components
*  Add `--manage <distro_name> --set-sparse` to wsl.exe
*  Fix various issues with the 'Update available' notification logic
*  Update WSLg to 1.0.57
    *  WSL GUI apps now have Windows snapping with the keyboard (Press WIN + an arrow key to snap to the side)
*  Update MSRDC to 1.2.4485

<hr /><em>This discussion was created from the release <a href='https://github.com/microsoft/WSL/releases/tag/2.0.0'>2.0.0</a>.</em>

**Comment by jamietre** [comment]
Hi! The link above: https://aka.ms/wslsep2023updatedocspage just redirects to bing.com.

Edit: Looks like it's already fixed, thank you!

**Reply by craigloewen-msft** [reply]
Haha yes you caught it right away while we were still making the docs PR public :)

**Comment by peteromio** [comment]
In the devblog article it is written "`autoMemoryReclaim=drop`", here "`autoMemoryReclaim=dropcache`" which would be correct?

**Reply by peteromio** [reply]
ty

**Reply by craigloewen-msft** [reply]
Fixed in the article!

**Comment by wr0nggg** [comment]
How do i revert this update? docker doesnt work

**Reply by jbdamiano** [reply]
docker service couldn't be started by systemd  on this version.
To rollback:

- uninstall Windows Subsystem for Linux ( in App menus)
- Remove all experimental feature from wslconfig
- on Powershell start:  `wsl --update --web-download`

**Reply by wr0nggg** [reply]
docker update to newest version works

**Reply by jbdamiano** [reply]
@wr0nggg What do you mean by "docker update to newest version". I'm WSL2 on Ubuntu 22.04 version 24.0.6-1~ubuntu.22.04~jammy. Based one the docker web site the 24.0.6 is the last version.

And after the WSL2 migration to 2.0.0 the docker.service and docker.stop failed to start by init.d

**Reply by wr0nggg** [reply]
docker desktop* sorry

**Reply by jbdamiano** [reply]
Ok I don't use docker-desktop. I used docker-ce and systemd to start the docker services.

**Reply by jbdamiano** [reply]
Issue https://github.com/microsoft/WSL/issues/10497, autoMemoryReclaim=dropcache works with docker-ce as well.

**Reply by cdccentral-sourcecontrol** [reply]
Docker daemon has this error “failed to start daemon: Devices cgroup isn’t mounted”

**Reply by Mr-Bridge-0** [reply]
@wr0nggg did you end up getting the 2.0.0 update to work with docker desktop?

**Reply by ZhangTianrong** [reply]
Append `cgroup2 /sys/fs/cgroup cgroup2 rw,nosuid,nodev,noexec,relatime,nsdelegate 0 0` to `/etc/fstab` and add `kernelCommandLine=cgroup_no_v1=all` to the `wsl2` section in `.wslconfig`. Then, `dockerd` seems to be starting correctly. This workaround was mentioned [here](https://github.com/microsoft/WSL/issues/9868#issuecomment-1490060424).

**Comment by daddycocoaman** [comment]
Hi. Pretty excited for this update, particularly the mirrored network. Thanks for the work. Is this update compatible with ARM64 builds?

```
psh ❯ type .wslconfig
[experimental]
autoMemoryReclaim=gradual
sparseVhd=true
networkingMode=mirrored
dnsTunneling=true
firewall=false
autoProxy=true

psh ❯ wsl --version
WSL version: 2.0.0.0
Kernel version: 5.15.123.1-1
WSLg version: 1.0.57
MSRDC version: 1.2.4485
Direct3D version: 1.608.2-61064218
DXCore version: 10.0.25880.1000-230602-1350.main
Windows version: 10.0.23545.1000

psh ❯ wsl
wsl: Hyper-V firewall is not supported
wsl: Mirrored networking mode is not supported, falling back to NAT networking
wsl: DNS Tunneling is not supported
```

**Reply by vasekboch** [reply]
The new experimental features requires Windows builds from Insider program. Check the docs page, they are marked with **

**Reply by daddycocoaman** [reply]
Insiders build --> `Windows version: 10.0.23545.1000`

Also, I should point out that this was after a `wsl --shutdown` and a reboot too.

**Reply by iosub** [reply]
Hi

Same error with the same windows version

**Reply by dducm** [reply]
Hello, same problem :

WSL version: 2.0.0.0
Kernel version: 5.15.123.1-1
WSLg version: 1.0.57
MSRDC version: 1.2.4485
Direct3D version: 1.608.2-61064218
DXCore version: 10.0.25880.1000-230602-1350.main
Windows version: 10.0.22631.2115

-> Hyper-V firewall is not supported
-> Mirrored networking mode is not supported, falling back to NAT networking

:(

**Reply by benhillis** [reply]
You need Windows version 25885 or later for the experimental networking features.

**Reply by daddycocoaman** [reply]
@benhillis That build number isn't on the chart (https://learn.microsoft.com/en-us/windows-insider/flight-hub/) but assuming that's true, then the features are only available in the Canary Channel? 😭

**Comment by iosub** [comment]
> Insiders build --> `Windows version: 10.0.23545.1000`
> 
> Also, I should point out that this was after a `wsl --shutdown` and a reboot too.
Hi

Same error with the same windows version

**Comment by flundstrom2** [comment]
The million dollar question:

Is the bug https://github.com/microsoft/WSL/issues/6982 which causes WSL to occupy 100% CPU finally fixed? 

If not, me and hundreds of others who have commented in that thread, are going to be very disappointed.

**Reply by stephanwehr** [reply]
For me, still 100% CPU usage after upgrade to version 2.

**Comment by niemsoen** [comment]
It's unclear to me, if I can use the networkMode=mirrored with Windows 10 22H2 if I am on a Windows Insider Program channel, or whether I have to be on Windows 11. In your [blog post](https://devblogs.microsoft.com/commandline/windows-subsystem-for-linux-september-2023-update/#new-networking-mode-mirrored) @craigloewen-msft  you say 

> this feature is currently only available to Windows Insiders canary and Release Preview Channel with the latest Windows 11, version 22H2

However in your [youtube tutorial](https://youtu.be/OB2Rfzy5V0Y?feature=shared&t=725) you say that `networkingMode=mirrored` is available on the Insiders Program, but don't specify the Windows version explicitly.

Does it work on Windows 10 Insider? Thank you so much for your reply :)

**Reply by craigloewen-msft** [reply]
Actually this is available only on the Windows Insiders canary channel and release preview channel right now. It will be going to Windows 11 22H2 eventually.

**Reply by crnolan** [reply]
Hi folks, bit late to this one, but does this mean there's no plan for a backport to Windows 10 for the new mirrored network mode?

**Comment by pleaselouis** [comment]
In my service program(session 0) execute command "wsl -l -v"
get response: Access is denied. Error code: Wsl/E_ACCESSDENIED 

my pc: win10 professionals 64-bit (10.0, Build 19045) (19041.vb_release.191206-1406)

WSL version： 2.0.0.0
kernel version： 5.15.123.1-1

But release note said:  Store WSL is now accessible from Session 0 (solves https://github.com/microsoft/WSL/issues/9231)
Does anyone know why??

---

update: in my win 11 pc , The result is the same

**Reply by craigloewen-msft** [reply]
Try running wsl.exe from this location: `C:\Program Files\WSL\wsl.exe` if you're running in session 0.

**Reply by pleaselouis** [reply]
I still got same error code. 
Access is denied. Error code: Wsl/E_ACCESSDENIED

//this is my c# code

        Process process = new Process();
        process.StartInfo.FileName = @"C:\Program Files\WSL\wsl.exe";
        process.StartInfo.Arguments = String.Format("-l -v");
        process.StartInfo.UseShellExecute = false;
        process.StartInfo.RedirectStandardOutput = true;
        process.StartInfo.RedirectStandardError = true;
        process.StartInfo.CreateNoWindow = true;

        process.StartInfo.StandardOutputEncoding = System.Text.Encoding.Unicode;

        process.Start();

        String result = process.StandardOutput.ReadToEnd();
        this.EventLog.WriteEntry(String.Format("cmd:{0} result:\n{1}\n", "wsl test", result));

        String error = process.StandardError.ReadToEnd();
        this.EventLog.WriteEntry(String.Format("cmd:{0} error:\n{1}\n", "wsl test", error));

**Reply by kingp0dd** [reply]
I am getting the same error even after running from `C:\Program Files\WSL\wsl.exe`

```
C:\Program Files\WSL>wsl.exe  --manage riserdev2 --set-sparse true
Conversion in progress, this may take a few minutes.
Access is denied.
Error code: Wsl/Service/E_ACCESSDENIED
```

**Comment by g2flyer** [comment]
Something in this update made reverting to an older version (1.3.11.0) failing: I get now an error "App installation failed with error message: error 0x80073D27: While processing the request, the system failed to register the windows.service extension due to the following error: The package installation failed because a version of the service exists outside of APPX ". Anybody has any idea how to overcome that error?

**Comment by ChGen** [comment]
When will be the next major public release of WSL2 in the MS Store?

**Comment by alvin-7** [comment]
The Wireguard cannot run as expected.

**Reply by nlvw** [reply]
In this case you probably need to run wireguard in Windows and not in your WSL instance.  I tested this with Cisco AnyConnect and it worked pretty well.

It sounds like the new mirror mode ties the WSL instance directly to the windows networking state/stack.  Which is desirable for most use cases but doesn't sound like it would be compatible with running a VPN inside of WSL itself.
